### PR TITLE
Fix some dashboard issues

### PR DESCRIPTION
**Pull request by somebody1234**
### Pull Request Description
Fixes some of #6662

Issues addressed:
- `ide watch` and `gui watch` should now use the desktop platform
- error screen should now be shown when passing invalid options
- password (both creating password when registering, and resetting password) should now warn on invalid input

### Screenshots
image

### Important Notes
Instead of checking whether `location.hostname === 'localhost'`, I've opted to use a constant defined by the build tool instead. This is to make it easier to merge the cloud IDE and desktop IDE entrypoints in the future, since it would be able to simply set `platform: Platform.cloud` in the build config.

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] The documentation has been updated, if necessary.
- [x] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [x] If GUI codebase was changed, the GUI was tested when built using `./run ide build`.

### CODE REVIEW

**Comment by somebody1234** [comment]
@wdanilo you might want to try out this PR?

File: app/ide-desktop/lib/client/watch.ts
Comment (wdanilo) [inline app/ide-desktop/lib/client/watch.ts]: 1. I believe this is bad solution. We have logic in Rust like "if the platform was desktop and frame is hidden, show window control buttons in WebGL". Basically, platform "web" means a web browser. When using `run ... watch` you can run the app either in electron or in the browser and we want to know where it is being used. This setting should reflect that. If you want to control deep links, I'd advise to check the url if it is localhost instead.

2. The problem is not only with watch scripts. If you install our app via installer, you run it, you should still be able to connect to it from chrome, as Electron serves that on localhost. This way of running should be supported correctly as well.

File: app/ide-desktop/lib/common/src/index.ts
Comment (wdanilo) [inline app/ide-desktop/lib/common/src/index.ts]: This is incompatible with Rust. In Rust we support platform "web", "macOS", "window", etc. We don't support "desktop", so this could throw strange things in Rust code sometimes.

File: app/ide-desktop/lib/content/bundle.ts
Comment (wdanilo) [inline app/ide-desktop/lib/content/bundle.ts]: copy-pasted code.

File: app/ide-desktop/lib/content/src/index.ts
Comment (wdanilo) [inline app/ide-desktop/lib/content/src/index.ts]: could you remove the HTML completely instead of hiding it? It helps with perf debugging of the app.

File: app/ide-desktop/lib/content/src/index.ts
Comment (wdanilo) [inline app/ide-desktop/lib/content/src/index.ts]: As I've written previously (I know this is an old comment), the comment is not true. Even if running in electron, you can connect to localhost via web browser, so the platform will be "web" then. So this is not true that this is desktop platform.

File: app/ide-desktop/lib/content/start.ts
Comment (wdanilo) [inline app/ide-desktop/lib/content/start.ts]: again copy-pasted code

File: app/ide-desktop/lib/content/watch.ts
Comment (wdanilo) [inline app/ide-desktop/lib/content/watch.ts]: another copy-pasted code

File: lib/rust/ensogl/pack/js/src/runner/index.ts
Comment (wdanilo) [inline lib/rust/ensogl/pack/js/src/runner/index.ts]: Is this code moved or changed as well?

**Review by wdanilo — CHANGES_REQUESTED** [review]

File: lib/rust/ensogl/pack/js/src/runner/index.ts
Comment (somebody1234) [reply lib/rust/ensogl/pack/js/src/runner/index.ts]: only moved. it needs to be accessed by `content/index.ts` so it shouldn't be an instance method. an alternative is just making it a static method, not sure whether it'd make sense for `content/index.ts` to do `app.App.showConfigOptions` though

File: app/ide-desktop/lib/common/src/index.ts
Comment (somebody1234) [reply app/ide-desktop/lib/common/src/index.ts]: this is for the dashboard only - the dashboard cannot access the IDE options, so it needs its own platform enum. maybe the naming can be improved though

File: lib/rust/ensogl/pack/js/src/runner/index.ts
Comment (wdanilo) [reply lib/rust/ensogl/pack/js/src/runner/index.ts]: The current solution is perfect. I asked only to know if I should review the content or not :)

File: lib/rust/ensogl/pack/js/src/runner/index.ts
Comment (wdanilo) [inline lib/rust/ensogl/pack/js/src/runner/index.ts]: Actually, all the code above seems very similar to what EnsoGL pack does. Instead of doing all the above, could we just call function from EnsoGL pack that checks the amount of unrecognized things, displays the config options if there are unrecognized flags, etc? I'd prefer to keep as much logic as possible in one place without duplicating it.

File: app/ide-desktop/lib/client/watch.ts
Comment (somebody1234) [reply app/ide-desktop/lib/client/watch.ts]: 1. i think the desktop IDE uses localhost too (it has a server, after all) so i'm not sure whether that will work for deciding whether deep links are supported. i guess it's a good solution for whether to show or hide the local backend/remote backend switcher though.
it will be incorrect when doing `npm run watch-dashboard` though, but that should be more or less fine because:
  a) it only happens during development, not when deployed to `cloud.enso.org` and
  b) `127.0.0.1` can be used for that instead
2. hmm... maybe we can check for `'isElectron' in window` instead? (and define `isElectron = true` in preload?)
  there would still be issues with `./run ide watch` though, when using the electron window, since that doesn't support deep links. not sure if there are any good solutions to that

File: lib/rust/ensogl/pack/js/src/runner/index.ts
Comment (somebody1234) [reply lib/rust/ensogl/pack/js/src/runner/index.ts]: that code is in `ensogl/pack`
the lines in `content/src/index.ts` are:
```js
        const unrecognized = contentConfig.OPTIONS.loadAll([app.urlParams()])
        if (unrecognized.length > 0) {
            app.showConfigOptions(contentConfig.OPTIONS, unrecognized)
        }
```
i think this is a reasonable amount of logic to have here, not sure if it can be reduced?
note that one is calling a method on `Group`, the other is calling a top-level function (previously in `App`)

File: app/ide-desktop/lib/common/src/index.ts
Comment (somebody1234) [reply app/ide-desktop/lib/common/src/index.ts]: maybe it should be removed and replaced with something like `supportsLocalBackend` instead?

**Comment by PabloBuchu** [comment]
@somebody1234 whats the status here? All the review comments are addressed? Can we solve conflicts so I could QA?

**Comment by wdanilo** [comment]
@PabloBuchu I don't believe my comments are applied, as I was not marked to re-review this PR.

**Comment by somebody1234** [comment]
not sure the new solution is any good either... but it should make more sense at least?

File: app/ide-desktop/lib/content/globals.d.ts
Comment (wdanilo) [inline app/ide-desktop/lib/content/globals.d.ts:11]: if this is true on macOS and Windows, we can compute it from the configuration, right? In the config, we have info what system it is running at - this is either "web" or macOS or windows - so instead of passing this flag, we could just compute that. Am I correct?

File: app/ide-desktop/lib/content/src/index.ts
Comment (wdanilo) [inline app/ide-desktop/lib/content/src/index.ts]: I think I was writing about it in my previous PR review. This logic is implemented in EnsoGL pack (if args are unrecognized, display the page with information about it). We can just use it from there, like `const parseOk = contentConfig.OPTIONS.loadAll([app.urlParams()])`, where `loadAll` will display this page if there were unrecognized args. I just don't like duplicating logic, so I'd prefer to keep such logic in one place.

File: app/ide-desktop/lib/content/src/index.ts
Comment (wdanilo) [reply app/ide-desktop/lib/content/src/index.ts]: This comment is not answered / fixed.

File: app/ide-desktop/lib/dashboard/src/authentication/src/authentication/cognito.ts
Comment (wdanilo) [inline app/ide-desktop/lib/dashboard/src/authentication/src/authentication/cognito.ts]: This string is used 3 times in this PR. It is error-prone, as doing typo in one place will not be visible. Such constants should be extracted as const values somewhere instead of hardcoding literals.

File: app/ide-desktop/lib/client/src/preload.ts
Comment (wdanilo) [inline app/ide-desktop/lib/client/src/preload.ts]: There is no info why this is needed. This makes the code hard to refactor, as the reader does not know how this is used and why it is required. As a reviewer, I also don't understand why this is required.

**Comment by wdanilo** [comment]
@somebody1234 I added a few comments. If you feel they are wrong/not needed or you'd like to chat, let's have a short call :)

File: app/ide-desktop/lib/content/globals.d.ts
Comment (somebody1234) [reply app/ide-desktop/lib/content/globals.d.ts:11]: i don't think so - we also need to know whether it's being run as `ide watch` or as `Enso.app`, since it's not possible to handle deep links on `ide watch`

File: app/ide-desktop/lib/content/src/index.ts
Comment (somebody1234) [reply app/ide-desktop/lib/content/src/index.ts]: ah, yeah. i figured  it was only three lines, but it's not a bad idea to make `loadAll` display the error dialog

**Comment by somebody1234** [comment]
the "this package is an Electron desktop app" comment has been removed without explanation as we are (currently) no longer using platform detection (for reasons explained in comments: `ide watch` does not count as "web" afaict, however it does not support deep links)

**Comment by somebody1234** [comment]
as for the other comments: most have been addressed
`IS_ELECTRON` has been replaced with detecting electron based on the user agent. the detection function comes with a small explanation, but more importantly it shouldn't ever break unless electron decides to change their user agent to something not including "electron"

File: app/ide-desktop/lib/content/globals.d.ts
Comment (wdanilo) [reply app/ide-desktop/lib/content/globals.d.ts:11]: I see that there is now word "installed" there, but I think adding explicit sentence that it does not work when running `ide watch` and similar things, would make things even more clear to a reader. Would you be OK with it?

File: app/ide-desktop/lib/content/src/index.ts
Comment (wdanilo) [inline app/ide-desktop/lib/content/src/index.ts]: Sorry for the comment to the same place as previously. Would you be OK with changing the name to `loadAllAndDisplayHelpIfUnsuccessful`? I know it's super long, but at least it tells exactly what it does. Currently, the name can be misleading.

**Review by wdanilo — APPROVED** [review]
Perfect! Let's do careful Q/A and let's merge it :)

**Comment by PabloBuchu** [comment]
QA issues:
Registration form doesn't work. I am typing my password `Korek12#$` and input is showing up:
[Here](https://stackoverflow.com/questions/58767980/aws-cognito-password-regex-specific-to-aws-cognito) is an example of valid cognito password regexp
Screenshot 2023-05-17 at 13 26 47

**Comment by somebody1234** [comment]
@PabloBuchu whoops, should be fixed now?

**Comment by PabloBuchu** [comment]
@somebody1234 if not now then when? ;) the problem clearly is with `PASSWORD_PATTERN` which is introduced in this PR

**Comment by PabloBuchu** [comment]
`run gui watch` and trying to sign in with github / google

Screenshot 2023-05-17 at 13 51 16

**Comment by PabloBuchu** [comment]
QA @somebody1234 :
1. build & run: 🟢 
2. debug scenes: 🟢 (even `ide` works fine now, great!. btw @wdanilo do we want (in separate issue) to add here `ide (deprecated)` or you wont need this entrypoint?)
3. Password registration password form not validating proper password 🔴 
4. `./run gui watch --skip-wasm-opt --skip-version-check`:
- login with github / google not working (see above comment with screenshot) 🔴 
- signing in with email/password 🟢 
- passing giberish parameters to search bar results wich `Unknown config option page` 🟢 
- when running with `authentication=True` and `featurePreview.newDashboard=True` and typing `?authentication=false&featurePreview.newDashboard=false` into search bar results with loading the old IDE 🟢 
- hot reload of rust files 🟢 
5. `/run ide watch --backend-source release --backend-release latest --skip-wasm-opt --skip-version-check --ide-option=-debug.dev-tools`:
- signing in with github / google I know deeplinking won't work and we act like normal browser with redirects but finally after signing in to github/google I am landing on `https://cloud.enso.org` 🔴 
- signing in with email / password 🟢 
- hot reload not working but it is also not working on develop so I will create separate issue for that 🟡

**Comment by somebody1234** [comment]
oh, nice catch. should the redirect with the oauth code be specialcased by content/index.ts?

**Comment by somebody1234** [comment]
3: i meant it should have been fixed by the new commit
5a: will fix soon
5c: it might have been broken during the switch to esbuild but I'm not too sure about that. i think the main issue with ide watch is that we are not using the esbuild server - i think it's possible to run an esbuild server just for the live reload notifications though

**Comment by PabloBuchu** [comment]
> oh, nice catch. should the redirect with the oauth code be specialcased by content/index.ts?

yeah I think so yeah.

**Comment by wdanilo** [comment]
@PabloBuchu what do you mean by `do we want (in separate issue) to add here ide (deprecated) or you wont need this entrypoint?)`? The `ide` entry point is extremely useful and we are using it all the time when developing Enso. I would even tell this is the most important debug scene of all of them. How is it deprecated? Where do you get this information from?

**Comment by PabloBuchu** [comment]
@wdanilo I just meant do we want to have entrypoint for old IDE. After we enable the newDashboard by default the `ide` entrypoint will lead to new dashboard.

I was just asking whether you will be using anyway the old one

**Comment by PabloBuchu** [comment]
QA looks 🟢 
- password input works fine now 🟢 
- google / github signing methods works both for `ide` and `gui` `watch` 🟢 
- still see some `Resources not found errors` when running `./run ide watch --backend-source release --backend-release latest --skip-wasm-opt --skip-version-check --ide-option=-debug.dev-tools` and `./dist/ide/mac-arm64/Enso.app/Contents/MacOS/Enso -debug.dev-tools -authentication -feature-preview.new-dashboard` 🟡 
Screenshot 2023-05-18 at 10 28 42


@wdanilo are you able to to give this PR a try and play around to see if anything is missing?

**Review by PabloBuchu — APPROVED** [review]

**Comment by somebody1234** [comment]
was attempting to figure out how `content` is being bundled into the final bundle, but still have no clue. @mwu-tow would you happen to know?

**Comment by somebody1234** [comment]
(it's possible (though unlikely?) that the `content/` bundle isn't being rebuilt, and so it's somehow using a `content/` bundle with `devMode` enabled)

**Comment by wdanilo** [comment]
I'll do the Q/A tomorrow. Sorry, I was unable to do it today :(

**Comment by wdanilo** [comment]
🔴🔴🔴 **Ok, I was playing with it. It works worse than previously. Previously I was able to start the app, now I am not able to make the app running at all.**

Also, I think we should not merge it until all build-related issues are fixed, which in this case means errors "resource not found". No one knows what such issues can cause on different platforms, and we should keep our build scripts free from errors.

I found the additional issues:
- After clicking "login with GitHub" and then pressing arrow back in the browser, the browser "hangs" - the tab loading spinner rotates forever until I click back again. The same happens with "login with google".
- After logging when opened in browser on localhost, I was asked to set username. When trying to do so, I got:
   image
- Very important regression - when trying to log in from electron, the login website was not opened in a pop-up, it was opened in the same electron window! (running as `./run ide watch --skip-version-check --skip-wasm-opt --backend-source release --backend-release latest`). This can't work this way even in development mode (in case it works differently when compiled as distribution), as the development mode should work as close as possible with dist to catch issues and check the behavior. But I haven't checked dist, so I don't know it this happens there as well.
- Another thing that was working better earlier. After logging with Google, and clicking on my phone "approve", in the past the App was redirected automatically, now I have to explicitly click a button in the app:
    image
- Actually, when trying to login from Electron (running as `./run ide watch --skip-version-check --skip-wasm-opt --backend-source release --backend-release latest`), I got the same error as above:
image

So basically, I can't log in and start the app.

**Comment by somebody1234** [comment]
- infinite loading: will investigate
- cannot set username on local backend: this is because the local backend is the default, however the remote backend is required for the authentication flow - this should be easy enough to fix
- regression re: the login website is opened inside the window - we currently use deep links to handle redirects, and that's only possible for installed apps, so it doesn't work for `ide watch` because that runs `electron` directly rather than `Enso.app`
  - if it (`ide watch` + third-party login) was working before that would be good to know, however afaict it didn't work before?
- having to click a button - i think this may be because google doesn't recognize electron?

**Comment by somebody1234** [comment]
re: login being opened in the same window - i have nothing against reverting, *however* that would mean (afaict) that it's impossible to test google and github logins using `ide watch`, since they wouldn't work at all (i may be wrong, but i haven't been able to; repro steps would be greatly appreciated)

*but* that does mean i'll hold off on reverting the google/github login stuff until it's confirmed we really don't want that behavior

**Comment by somebody1234** [comment]
infinite loading: can't repro :/

**Comment by wdanilo** [comment]
Guys, if this is working now, let's merge it and let's fix following issues in the next PRs.

**Comment by somebody1234** [comment]
@wdanilo i'm not sure all the issues from the QA have been resolved - as mentioned above:
- some i cannot repro
- some are intentionally like that
- some have been fixed

**Comment by wdanilo** [comment]
@somebody1234 I've read your comments above, sorry for not being clear enough. I understand that some of them you can't repro, let's skip them for now. Regarding the ones that are designed this way – I would like to talk more about it, but probably not now, as these are not blockers. Regarding the fixes - they fixed the most important things, so let's merge it and continue improvements in next PRs :) 

The most important thing is that everything should work with `./run watch`. If that works, we are at home :)

**Comment by somebody1234** [comment]
"resource not found" errors i *think* originated either from a `watch-dashboard` or `gui` tab that was still open, i *think* at no point was devMode set to true for `ide watch` - and *definitely* not for `dist`. however if it can be reliably repro'd it should be simple enough to pinpoint the problem

**Comment by somebody1234** [comment]
but in that case - i think i'll do a quick QA just to make sure everything's good before marking as ready to merge

**Comment by wdanilo** [comment]
@somebody1234 good idea. Lets do it :)

**Comment by somebody1234** [comment]
QA:
`npm run watch-dashboard`:
- :warning: cloud signup works: signup link contains `redirect_url=enso://auth/confirmation`. this is temporarily fixed, but backend handling should be fixed.
  - can repro the toast notification appearing in the middle on `cloud.enso.org`. i think rather than appearing in the middle, it's just appearing in the top right of the space below the login form. either way, not sure it's worth spending the time to fix it because it

`./run ide watch`:
:heavy_check_mark: after fix, no `redirect_url=enso://`
:heavy_check_mark: login works
:heavy_check_mark: no asset fetch errors
:heavy_check_mark: full registration flow (email) works; dashboard is initially set to local backend
:heavy_check_mark: backend switcher works
:heavy_check_mark: `localhost:8080` works, backend switcher works and local backend works
:heavy_check_mark: can edit local project on browser

`./run ide build`:
:heavy_check_mark: after fix, there is still `redirect_url=enso://`
:heavy_check_mark: login works
:heavy_check_mark: no asset fetch errors
:heavy_check_mark: backend switcher works
:heavy_check_mark: `localhost:8080` works - was :x: broken because `supportsDeepLinks` was unconditionally true; now fixed in dashboard module.
:heavy_check_mark: can edit local project on browser
  :question: weird error: `app/gui/suggestion-database/src/lib.rs:460 Error when applying update for entry 13: InvalidArgumentIndex { name: "operator9", index: 0 }`. pretty sure it's unrelated to this PR though.
  :question: shows red and green buttons. i think this is a non-issue; quite odd because `startup.platform` should be the default (`web` because it is not set by anything)

`./run gui watch`:
:heavy_check_mark: login works
:heavy_check_mark: "please register first" works
:heavy_check_mark: no asset fetch errors
:heavy_check_mark: backend switcher works
:heavy_check_mark: can edit local project
:heavy_check_mark: shows error message when project manager is not running. previously :x: no error message and :x: project manager didn't stop trying to connect. project manager reconnection logic also changed to stop after 10 seconds instead of 10 tries.

:heavy_check_mark: error screen on passing invalid options
:heavy_check_mark: password validation prevents sending form. previously :warning: working, but logging warnings on chrome

**Comment by wdanilo** [comment]
Ok, @PabloBuchu could you give a final green light for merging it now, please? :)

**Comment by wdanilo** [comment]
@somebody1234 oh wow, that is a detailed information, thank you! In such a case, let's merge it and lets fix the following issues in next PRs :)

**Pull request merged**
